paperqa/llms.py

```diff
@@ -149,6 +152,40 @@ async def embed_documents(self, client, texts):
         return np.concatenate(all_embeds, axis=1)
 
 
+class VoyageAIEmbeddingModel(EmbeddingModel):
+    """A wrapper around Voyage AI's client lib."""
+
+    name: str = Field(default="voyage-large-2")
+    embedding_type: str = Field(default="document")
+    batch_size: int = 10
+
+    def set_mode(self, mode: str):
+        if mode == "query":
+            self.embedding_type = "query"
+        elif mode == "document":
+            self.embedding_type = "document"
+        else:
+            raise NotImplementedError(
+                f"Mode {mode} is not implemented for VoyageAI embeddings."
+            )
+
+    async def embed_documents(self, client: Any, texts: list[str]) -> list[list[float]]:
+        if client is None:
+            raise ValueError(
+                "Your client is None - did you forget to set it after pickling?"
+            )
```

Comment:
Do we really need this? Imo this is the problem `mypy`/typing is supposed to fix

Reply:
this is a mutable state thing where the client gets wiped every time you pickle it, and you manually need to reset it.